apps/performance-tests/src/util/BlockHandler.ts

```diff
@@ -0,0 +1,76 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import blocked from "blocked";
+import { SortedArray } from "@itwin/core-bentley";
+
+export interface Summary {
+  count: number;
+  max?: number;
+  p95?: number;
+  median?: number;
+}
+
+/**
+ * This class measures the durations of time when main thread is blocked.
+ * This is measured by running a timer which detects cases when it is fired later than expected.
+ */
+export class BlockHandler {
+  private readonly _samples = new SortedArray<number>((a, b) => a - b);
+  private _timer?: NodeJS.Timer;
+
+  public getSummary(): Summary {
+    const arr = this._samples.extractArray();
+    const count = arr.length;
+    const max = count ? arr[count - 1] : undefined;
+    const p95 = getP95(arr);
+    const median = getMedian(arr);
+    return {
+      count,
+      max,
+      p95,
+      median,
+    };
+  }
+
+  /**
+   * Starts the timer and records instances of abnormally long blocking.
+   * @param threshold The minimum amount of blocking that will be considered abnormal.
+   * @param interval Delay in time between each blocking check.
+   */
+  public start(threshold: number = 20, interval: number = 10) {
+    this._samples.clear();
+    this._timer = blocked((time) => this._samples.insert(time), { threshold, interval });
+  }
+
+  /** Stops the blocking timer. */
+  public stop() {
+    clearTimeout(this._timer);
+  }
+}
+
+function getP95(arr: number[]): number | undefined {
+  if (arr.length <= 1) {
+    return undefined;
+  }
+
+  const length95 = Math.ceil(0.95 * arr.length);
+  let sum95 = 0;
+  for (let i = 0; i < length95; ++i) {
+    sum95 += arr[i];
+  }
+  return sum95 / length95;
+}
```

Comment:
we want 95th percentile rather than its average (similar to median, which is the same as p50)

Reply:
I am not an expert on statistics 😄 
I hope this is ok now